### PR TITLE
IO module error types cleanup

### PR DESCRIPTION
**Pull request by jeremiah-corrado**
This PR makes several changes to the IO module:

- Corrects the error types thrown by a few methods
    - For example, `readAll(type t = string)` now throws an `EofError` when nothing is read (because the ``fileReader`` is already at EOF). This is meant to be analogous to the behavior of the `readAll(s: string): int` overload, which returns `0` in such a case.
    - Most of the other changes were initiated by a bug fix in `_qio_buffered_write` (in `qio.c`). This function is now aware that a writing channel can have an upper bound (`ch->end_pos`), and will no longer attempt to write past it.
- Fixes some cases where error types were documented incorrectly
    - For example, `UnexpectedEofError` was removed from the docs for `readBinary(string)` because this method is intended to read until EOF, so `UnexpectedEofError` should never be thrown.
- Some general wordsmithing an consistency improvements in the docs
- Adds a subsection to the "Error Handling" section about what `SystemError` means in the IO module
- Fixes a bug where a `fileReader`/`fileWriter`'s region upper bound could overflow during initialization
- Updates a variety of tests to reflect error type changes, and to cover some previously untested cases

Testing:
- [x] inspected built docs
- [x] paratest
- [x] check Arkouda's use of modified error types

### CODE REVIEW

File: test/library/standard/IO/readAll/readBytesType.chpl [hunk 1]
```diff
@@ -6,3 +6,10 @@ ch.close();
 
 writeln(b);
 writeln(b.type == bytes);
+
+// nothing remaining to read in the channel:
+try {
+    ch.readAll(bytes);
+} catch e {
+    writeln(e);
+}
```
Comment (lydia-duncan) [inline test/library/standard/IO/readAll/readBytesType.chpl:15]: Missing newline at end of file

File: test/library/standard/IO/readAll/readStringType.chpl [hunk 2]
```diff
@@ -6,3 +6,10 @@ ch.close();
 
 writeln(s);
 writeln(s.type == string);
+
+// nothing remaining to read in the channel:
+try {
+    ch.readAll(bytes);
+} catch e {
+    writeln(e);
+}
```
Comment (lydia-duncan) [inline test/library/standard/IO/readAll/readStringType.chpl:15]: Ditto missing newline here

**Review by lydia-duncan — COMMENTED** [review]
This is a lot of good work!  My comments are mostly minor adjustments

File: modules/standard/IO.chpl [hunk 3]
```diff
@@ -4501,7 +4529,7 @@ proc file.lines(param locking:bool = true, start:int(64) = 0,
 
 /* Used to control the behavior of the region argument for :proc:`file.lines`.
    When set to ``true``, the region argument will fully specify the bounds that
-   this function would cover.  When set to ``false``, the region argument will
+   this function would cover.  When set to ``false``, the region argument useNewFileWriterRegionBounds
```
[the remark this answers is not in the document]
Comment (jeremiah-corrado) [reply modules/standard/IO.chpl:4532]: oops

File: modules/standard/IO.chpl [hunk 4]
```diff
@@ -6023,33 +6052,39 @@ proc fileReader.readline(arg: [] uint(8), out numRead : int, start = arg.domain.
...
-            returns the number of array elements that were set by this call.
-
-  :throws BadFormatError: Thrown if the line is longer than `maxSize`. It leaves
-                          the fileReader offset at the beginning of the
-                          offending line.
-  :throws SystemError: Thrown if data could not be read from the fileReader for
-                       another reason.
+  :arg stripNewline: Whether to strip the trailing ``\n`` from the line. If
+                     ``true``, the newline isn't counted in the number of
+                     bytes read.
```
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply modules/standard/IO.chpl:6071]: True true.  I'm probably over thinking it

**Comment by jeremiah-corrado** [comment]
Thanks for the review @lydia-duncan! I believe everything has been addressed.

**Review by lydia-duncan — APPROVED** [review]
Nice!  Assuming you've re-run testing for the test changes, I think this is good to go!

**Pull request merged**